datastore/postgres/deletemanifests.go

```diff
@@ -33,55 +36,58 @@ var (
-	rm, err := s.deleteManifests(ctx, d)
-	if err != nil {
-		return nil, err
-	}
-	return rm, s.layerCleanup(ctx)
+	return s.deleteManifests(ctx, d)
 }
 
-func (s *IndexerStore) deleteManifests(ctx context.Context, d []claircore.Digest) ([]claircore.Digest, error) {
-	const deleteManifest = `DELETE FROM manifest WHERE hash = ANY($1::TEXT[]) RETURNING manifest.hash;`
-	var err error
-	defer promTimer(deleteManifestsDuration, "deleteManifest", &err)()
-	defer func(e *error) {
-		deleteManifestsCounter.WithLabelValues("deleteManifest", success(*e)).Inc()
-	}(&err)
-	rows, err := s.pool.Query(ctx, deleteManifest, digestSlice(d))
-	if err != nil {
-		return nil, err
-	}
-	defer rows.Close()
-	rm := make([]claircore.Digest, 0, len(d)) // May over-allocate, but at least it's only doing it once.
-	for rows.Next() {
-		i := len(rm)
-		rm = rm[:i+1]
-		err = rows.Scan(&rm[i])
-		if err != nil {
-			return nil, err
+func (s *IndexerStore) deleteManifests(ctx context.Context, ds []claircore.Digest) ([]claircore.Digest, error) {
+	const (
+		getManifestID  = `SELECT id FROM manifest WHERE hash = $1`
+		getLayers      = `SELECT layer_id FROM manifest_layer WHERE manifest_id = $1;`
+		deleteManifest = `DELETE FROM manifest WHERE id = $1;`
+		deleteLayers   = `DELETE FROM layer WHERE id = ANY($1);`
+	)
+
+	deletedManifests := make([]claircore.Digest, 0, len(ds))
+	for _, d := range ds {
+		// Get manifest ID
+		var manifestID int64
+		err := s.pool.QueryRow(ctx, getManifestID, d).Scan(&manifestID)
+		switch {
+		case errors.Is(err, nil):
+		case errors.Is(err, pgx.ErrNoRows):
+			continue
+		default:
+			return nil, fmt.Errorf("unable query manifest: %w", err)
 		}
-	}
-	err = rows.Err()
-	if err != nil {
-		return nil, err
-	}
-	zlog.Debug(ctx).
-		Int("count", len(rm)).
-		Int("nonexistant", len(d)-len(rm)).
-		Msg("deleted manifests")
-	return rm, nil
-}
 
-func (s *IndexerStore) layerCleanup(ctx context.Context) (err error) {
-	const layerCleanup = `DELETE FROM layer WHERE NOT EXISTS (SELECT FROM manifest_layer WHERE manifest_layer.layer_id = layer.id);`
-	defer promTimer(deleteManifestsDuration, "layerCleanup", &err)()
-	tag, err := s.pool.Exec(ctx, layerCleanup)
-	deleteManifestsCounter.WithLabelValues("layerCleanup", success(err)).Inc()
-	if err != nil {
-		return err
+		// Get all layer IDs
+		lrows, err := s.pool.Query(ctx, getLayers, manifestID)
+		if err != nil {
+			return nil, fmt.Errorf("unable to query layers: %w", err)
+		}
+		defer lrows.Close() // TODO: revisit stacking up defers
+		lIDs := []int64{}
+		for lrows.Next() {
+			var layerID int64
+			err = lrows.Scan(&layerID)
+			if err != nil {
+				return nil, fmt.Errorf("unable to scan layer ID: %w", err)
+			}
+			lIDs = append(lIDs, layerID)
+		}
+		// TODO: feedback for how things went
+		// Delete manifest
+		_, err = s.pool.Exec(ctx, deleteManifest, manifestID)
+		if err != nil {
+			return nil, fmt.Errorf("unable to delete manifest: %w", err)
+		}
+		// Delete layers
```

Comment:
This is actually wrong, these layers could also be associated with other manifests.